pandas/core/groupby/ops.py

```diff
@@ -261,7 +274,15 @@ def get_out_dtype(self, dtype: np.dtype) -> np.dtype:
                 out_dtype = "object"
         return np.dtype(out_dtype)
 
-    def get_result_dtype(self, dtype: DtypeObj) -> DtypeObj:
+    @overload
+    def _get_result_dtype(self, dtype: np.dtype) -> np.dtype:
+        ...
+
+    @overload
+    def _get_result_dtype(self, dtype: ExtensionDtype) -> ExtensionDtype:
```

Comment:
Would a TypeVar work here? Might end up being useful elsewhere too

Reply:
i dont think that would be accurate here

Reply:
Might be misunderstanding overloads, but is this expressing `ExtensionDtype -> ExtensionDtype` and `np.dtype -> np.dtype`? 

So was thinking a TypeVar like `TypeVar('DtypeObjT', np.dtype, ExtensionDtype)`

Reply:
do TypeVars not preserve subclasses?

Reply:
Running `mypy` on something like

```
from typing import TypeVar

AnyStr = TypeVar('AnyStr', str, bytes)


class Upper(str):
    def __new__(cls, val):
        return str.__new__(cls, val.upper())


class Lower(str):
    def __new__(cls, val):
        return str.__new__(cls, val.lower())


def swap_lower_upper(s: AnyStr) -> AnyStr:
    if isinstance(s, bytes):
        return s
    elif isinstance(s, Lower):
        return Upper(s)
    else:
        return Lower(s)

swap_lower_upper(Lower("hello"))
swap_lower_upper(Upper("hello"))
```
gives no error, so I don't think subclasses are preserved.

Reply:
cool.  will make a separate PR to implement DtypeObjT

Reply:
implenting this with DtypeObjT and getting rid of the overloads im seeing

```
error: Incompatible return value type (got "dtype[signedinteger[_64Bit]]", expected "ExtensionDtype")  [return-value]
```

Reply:
Will have time to look in more detail later, initial guess would be that this return is the problem https://github.com/pandas-dev/pandas/blob/59b2db14f9c47a1b34da9efc12d708d43bcc893f/pandas/core/groupby/ops.py#L288-L290

IIRC mypy can't narrow based on type equality, so it's thinking that an `ExtensionDtype` can hit that return. But then not sure why it wouldn't complain with the overload, but those are confusing :)